In [13]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

In [14]:
df = pd.read_csv('E:/postgre/credit_risk.csv', header = 0)

In [15]:
df.head()

Id  Age  Income      Home  Emp_length     Intent  Amount   Rate  Status  \
0   0   22   59000      RENT       123.0   PERSONAL   35000  16.02       1   
1   1   21    9600       OWN         5.0  EDUCATION    1000  11.14       0   
2   2   25    9600  MORTGAGE         1.0    MEDICAL    5500  12.87       1   
3   3   23   65500      RENT         4.0    MEDICAL   35000  15.23       1   
4   4   24   54400      RENT         8.0    MEDICAL   35000  14.27       1   

   Percent_income Default  Cred_length  
0            0.59       Y            3  
1            0.10       N            2  
2            0.57       N            3  
3            0.53       N            2  
4            0.55       Y            4

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              32581 non-null  int64  
 1   Age             32581 non-null  int64  
 2   Income          32581 non-null  int64  
 3   Home            32581 non-null  object 
 4   Emp_length      31686 non-null  float64
 5   Intent          32581 non-null  object 
 6   Amount          32581 non-null  int64  
 7   Rate            29465 non-null  float64
 8   Status          32581 non-null  int64  
 9   Percent_income  32581 non-null  float64
 10  Default         32581 non-null  object 
 11  Cred_length     32581 non-null  int64  
dtypes: float64(3), int64(6), object(3)
memory usage: 3.0+ MB


**Null Value Imputation**

In [17]:
df.Emp_length = df.Emp_length.fillna(df.Emp_length.mean())
df.Rate = df.Rate.fillna(df.Rate.mean())

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32581 entries, 0 to 32580
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              32581 non-null  int64  
 1   Age             32581 non-null  int64  
 2   Income          32581 non-null  int64  
 3   Home            32581 non-null  object 
 4   Emp_length      32581 non-null  float64
 5   Intent          32581 non-null  object 
 6   Amount          32581 non-null  int64  
 7   Rate            32581 non-null  float64
 8   Status          32581 non-null  int64  
 9   Percent_income  32581 non-null  float64
 10  Default         32581 non-null  object 
 11  Cred_length     32581 non-null  int64  
dtypes: float64(3), int64(6), object(3)
memory usage: 3.0+ MB


**Dummy Value creation**

In [19]:
df = pd.get_dummies(df, columns = ['Home','Default','Intent'], drop_first = True)

In [20]:
df.head()

Id  Age  Income  Emp_length  Amount   Rate  Status  Percent_income  \
0   0   22   59000       123.0   35000  16.02       1            0.59   
1   1   21    9600         5.0    1000  11.14       0            0.10   
2   2   25    9600         1.0    5500  12.87       1            0.57   
3   3   23   65500         4.0   35000  15.23       1            0.53   
4   4   24   54400         8.0   35000  14.27       1            0.55   

   Cred_length  Home_OTHER  Home_OWN  Home_RENT  Default_Y  Intent_EDUCATION  \
0            3           0         0          1          1                 0   
1            2           0         1          0          0                 1   
2            3           0         0          0          0                 0   
3            2           0         0          1          0                 0   
4            4           0         0          1          1                 0   

   Intent_HOMEIMPROVEMENT  Intent_MEDICAL  Intent_PERSONAL  Intent_VENTURE  
0                       0               0                1               0  
1                       0               0                0               0  
2                       0               1                0               0  
3                       0               1                0               0  
4                       0               1                0               0

## Classification using random forest

In [12]:
from sklearn.ensemble import RandomForestClassifier

X-Y Split

In [21]:
x = df.loc[:,df.columns!="Default_Y"]
type(x)

pandas.core.frame.DataFrame

In [22]:
y = df["Default_Y"]

In [23]:
y.head()

0    1
1    0
2    0
3    0
4    1
Name: Default_Y, dtype: uint8

Test-Train Split

In [25]:
from sklearn.model_selection import train_test_split

In [26]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

Fitting the Model

In [27]:
rf_clf = RandomForestClassifier(n_estimators=1000, n_jobs=-1, random_state=42)

In [28]:
rf_clf.fit(x_train, y_train)

RandomForestClassifier(n_estimators=1000, n_jobs=-1, random_state=42)

In [30]:
from sklearn.metrics import accuracy_score, confusion_matrix

In [31]:
confusion_matrix(y_test, rf_clf.predict(x_test))

array([[4898,  503],
       [ 643,  473]], dtype=int64)

In [34]:
y_train_pred = rf_clf.predict(x_train)
y_test_pred = rf_clf.predict(x_test)

In [35]:
accuracy_score(y_test, y_test_pred)

0.8241522172778886

The accuracy score is 82%. Giving the model a high level of accuracy.